// @therootnetwork/api-types@0.7.2/dist/interfaces/definitions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.syloDataPermissions = exports.sft = exports.assetsExt = exports.xrplBridge = exports.runtime = exports.nft = exports.ethy = exports.ethWallet = exports.dex = void 0;
var definitions_1 = require("./dex/definitions");
Object.defineProperty(exports, "dex", { enumerable: true, get: function () { return __importDefault(definitions_1).default; } });
var definitions_2 = require("./ethWallet/definitions");
Object.defineProperty(exports, "ethWallet", { enumerable: true, get: function () { return __importDefault(definitions_2).default; } });
var definitions_3 = require("./ethy/definitions");
Object.defineProperty(exports, "ethy", { enumerable: true, get: function () { return __importDefault(definitions_3).default; } });
var definitions_4 = require("./nft/definitions");
Object.defineProperty(exports, "nft", { enumerable: true, get: function () { return __importDefault(definitions_4).default; } });
var definitions_5 = require("./runtime/definitions");
Object.defineProperty(exports, "runtime", { enumerable: true, get: function () { return __importDefault(definitions_5).default; } });
var definitions_6 = require("./xrplBridge/definitions");
Object.defineProperty(exports, "xrplBridge", { enumerable: true, get: function () { return __importDefault(definitions_6).default; } });
var definitions_7 = require("./assetsExt/definitions");
Object.defineProperty(exports, "assetsExt", { enumerable: true, get: function () { return __importDefault(definitions_7).default; } });
var definitions_8 = require("./sft/definitions");
Object.defineProperty(exports, "sft", { enumerable: true, get: function () { return __importDefault(definitions_8).default; } });
var definitions_9 = require("./syloDataPermissions/definitions");
Object.defineProperty(exports, "syloDataPermissions", { enumerable: true, get: function () { return __importDefault(definitions_9).default; } });
